controllers/volumereplicationgroup_controller.go

```diff
@@ -409,94 +420,231 @@ func (v *VRGInstance) processAsPrimary() (ctrl.Result, error) {
-// metadata in separate functions and calling them separately. In future,
-// if creation of VolumeReplication CR and backing up of pv metadata should
-// happen together (i.e for each pvc create VolumeReplication CR and then
-// backup corresponding PV metadata), then it can be put in a single function.
-func (v *VRGInstance) handlePersistentVolumeClaims() bool {
-	requeue := true
-
-	if requeueResult := v.createVolumeReplicationResources(); requeueResult {
-		v.log.Info("Requeuing resource")
+// preparePVCForVRProtection processes prerequisites of any PVC that needs VR protection
+func (v *VRGInstance) preparePVCForVRProtection(pvc *corev1.PersistentVolumeClaim,
+	log logr.Logger) bool {
+	const requeue bool = true
+
+	// if the PVC is not bound yet, don't proceed
+	if pvc.Status.Phase != corev1.ClaimBound {
+		log.Info("Requeuing as PersistentVolumeClaim is not bound", "pvcPhase", pvc.Status.Phase)
 
 		return requeue
 	}
 
-	if requeueResult := v.handlePersistentVolumes(); requeueResult {
-		v.log.Info("Requeuing resource")
+	// Check if PVC is already protected
+	pvcProtected, err := v.ensurePVCVRProtected(pvc, log)
+	if err != nil {
+		log.Info("Requeuing as determining PersistentVolumeClaim protection failed", "errorValue", err)
+
+		return requeue
+	}
+
+	if pvcProtected {
+		return !requeue
+	}
+
+	// Add `vr-protecting` finalizer to PVC
+	if err = v.addProtectingFinalizerToPVC(pvc, log); err != nil {
+		log.Info("Requeuing as protecting PersistentVolumeClaim failed", "errorValue", err)
+
+		return requeue
+	}
+
+	// Change PV `reclaimPolicy` to "Retain"
+	if err = v.retainPVForPVC(*pvc, log); err != nil {
+		log.Info("Requeuing as retaining PersistentVolume failed", "errorValue", err)
 
 		return requeue
 	}
 
 	return !requeue
 }
 
-func (v *VRGInstance) createVolumeReplicationResources() bool {
-	requeue := false
+// retainPVForPVC updates the PV reclaim policy to retain for a given PVC
+func (v *VRGInstance) retainPVForPVC(pvc corev1.PersistentVolumeClaim, log logr.Logger) error {
+	// Get PV bound to PVC
+	pv := &corev1.PersistentVolume{}
+	pvObjectKey := client.ObjectKey{
+		Name: pvc.Spec.VolumeName,
+	}
 
-	for _, pvc := range v.pvcList.Items {
-		pvcNamespacedName := types.NamespacedName{Name: pvc.Name, Namespace: pvc.Namespace}
-		log := v.log.WithValues("pvc", pvcNamespacedName.String())
+	if err := v.reconciler.Get(v.ctx, pvObjectKey, pv); err != nil {
+		log.Error(err, "Failed to get PersistentVolume", "volumeName", pvc.Spec.VolumeName)
 
-		// if the PVC is not bound yet, dont proceed.
-		if pvc.Status.Phase != corev1.ClaimBound {
-			log.Info("Requeuing as PersistentVolumeClaim is not bound",
-				"pvcPhase", pvc.Status.Phase)
+		return fmt.Errorf("failed to get PersistentVolume resource (%s) for"+
+			" PersistentVolumeClaim resource (%s/%s) belonging to VolumeReplicationGroup (%s/%s), %w",
+			pvc.Spec.VolumeName, pvc.Name, pvc.Namespace, v.instance.Name, v.instance.Namespace, err)
```

Comment:
(nit), do we need to log the v.intance.Namespace? Since we get the namespace from pvc.Namespace, probably just logging v.instance.Name is sufficient? I am fine with the existing way though. Just a thought.

Reply:
I was attempting to be future proof, in case we deal with different namespaces... I would like to leave it as is, unless strongly preferred otherwise.